### DISCUSSION TITLE
How to parse attribute on loading from a database

### DISCUSSION THREAD

**Discussion by tokarenko** [sqlalchemy/sqlalchemy#10377] [Usage Questions] [answered]
Please advise how to attach parser to an attribute that is transparently invoked when the attribute is modified or read from database? By transparent parser invocation I mean that it generates parsed value for internal use only, keeping attribute value and behavior intact. SQLAlchemy validator suggested in https://github.com/sqlalchemy/sqlalchemy/discussions/10295 works fine on object creation and modification in memory but is not invoked when object is read from a database:
```
@mapper_registry.mapped
@dataclass
class CollectionPolicy:
    id: int = field(init=False)
    collection_object_filter: str

    @validates("collection_object_filter")
    def _parse_collection_object_filter(self, key, value):
        self._parsed_filter = parse(value)
        return value

    __table__ = Table(
        "collection_policy",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("collection_object_filter", Text),
    )
```

dataclasses property decorator does not work also.
post_init and setter are not called when the object is loaded from a database:

```
@mapper_registry.mapped
@dataclass
class CollectionPolicy:
    id: int = field(init=False)
    collection_object_filter: InitVar[str]

    def __post_init__(self, collection_object_filter):
        self.collection_object_filter = collection_object_filter

    @property
    def collection_object_filter(self) -> str:
        return self._filter

    @collection_object_filter.setter
    def collection_object_filter(self, value) -> None:
        self._filter = value
        self._parsed_filter = parse(value)

    __table__ = Table(
        "collection_policy",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("collection_object_filter", Text),
    )

    __table__ = Table(
        "collection_policy",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("collection_object_filter", Text),
    )

    __mapper_args__ = {
        "properties": {
            "_filter": __table__.c.collection_object_filter
        }
    }
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

You can probably look at the instance events. This seem to be appropriate for your use case https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.load

**Reply by tokarenko** [reply]
Thank you @CaselIT for your suggestion. I explored it further and found [reconstructor](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.reconstructor) decorator. It provides for validator (as a parser) invocation on load from a database. Use of reconstructor is a good fit in my case because there is inheritance from the class and event handlers need to be registered automatically for inheriting classes.
```
@mapper_registry.mapped
@dataclass
class CollectionPolicy:
    id: int = field(init=False)
    collection_object_filter: str

    @validates("collection_object_filter")
    def _parse_collection_object_filter(self, key, value):
        self._parsed_filter = parse(value)
        return value

    @reconstructor
    def _sqlalchemy_reconstructor(self):
        self._parse_collection_object_filter(None, self.collection_object_filter)

    __table__ = Table(
        "collection_policy",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("collection_object_filter", Text),
    )
```

**Comment by zzzeek** [comment]
if ORM hooks are still too specific, the general case of database types that need to be marshalled to/from a Python object value is handled by types, see https://docs.sqlalchemy.org/en/20/core/custom_types.html#augmenting-existing-types

**Reply by tokarenko** [reply]
Thank you @zzzeek for your quick reply and the suggestion! I was also thinking about resorting to custom types in the absence of other options. I am glad there are https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.reconstructor that fits good for my case with class inheritance.
